CitizenLabDotCo/citizenlab
# [CL-2746] Add content_builder layouts & images to serializer

## Summary

Adds `content_builder_layouts` & `content_builder_layout_images` to template serializer.rb, and ensures a template containing representations of `content_builder_layout_images` can be applied.

Please see [comment](https://github.com/CitizenLabDotCo/citizenlab/pull/3887#discussion_r1096561344
) explaining need to remove validation of presence of `image` in `layout_image` model.

## Testing

Confirmed working with local manual testing

<details>
  <summary>Click for details of testing method</summary>

1. Create source tenant locally using local AdminHQ + base template 
2. Edit source tenant to taste (delete projects/add projects, etc.)
3. Open shell in `citizenlab_web` container, with `docker exec -it <container_ID> /bin/bash`
4. Export template of source tenant with `bundle exec rake templates:export[‘source.tenant.hostname.’, ‘template_filename.yml’] ` Template file appears in `/back` folder. If the template contains references to urls that include `localhost:3000`, then replace this part of the urls with `localhost:4000`
5. Open rails console in `citizenlab_web` container with `rails c`
6. FactoryBot.create(:tenant) (repeat until it works, if you’ve already created tenants with this command)
7. _.switch! to switch to new tenant
8. run `::MultiTenancy::TenantTemplateService.new.resolve_and_apply_template YAML.load(open('template_filename.yml').read)` [Actually, _**far quicker**_ is to remove the tagged logging in `TenantTemplateService` `resolve_and_apply_template method` & use the `templates:import` task & kill logging in the rake task, but harder to describe here]
9.  Examine the db of the newly created tenant and layout_image files created in tenant's directory in /back/public/uploads, to confirm `content_builder_layouts` & `content_builder_layout_images` have been created in the new tenant.

To view the new layouts on the FE:

10. Use AdminHQ (locally) 'Make me an admin' to make the copied user an admin on the newly created tenant.
11. set `OVERRIDE_HOST=hostname.of.new.tenant in back-safe.env` and restart docker containers with `ctrl-C`, `docker-compose up`
12. Open rails console in `citizenlab_web` container, with `docker exec -it <container_ID> rails c`
13. Add login and content_builder settings with:
```Ruby
[1] pry(main)> Apartment::Tenant.switch! 'tenant-6_citizenlab_co'
=> nil
[2] pry(main)> cfg = AppConfiguration.first
…
[3] pry(main)> cfg.settings['password_login'] = {"phone": false, "allowed": true, "enabled": true, "enable_signup": true, "minimum_length": 8, "phone_email_pattern": "phone+__PHONE__@test.com"}
=> {:phone=>false,
 :allowed=>true,
 :enabled=>true,
 :enable_signup=>true,
 :minimum_length=>8,
 :phone_email_pattern=>"phone+__PHONE__@test.com"}
[4] pry(main)> cfg.settings['content_builder'] = {"allowed": true, "enabled": true}
=> {:allowed=>true, :enabled=>true}
[5] pry(main)> cfg.save!
…
2023-02-04 16:36:34.862038 D [101:7740 log_subscriber.rb:126] ActiveRecord::Base --   TRANSACTION (1.1ms)  COMMIT
=> true
```
14. Refresh page, sign up &/or log in, and examine project description layouts and the display of the project description (using the layouts). Confirm layouts are equivalent to those of source tenant, including the display of copied images.
</details>
